Basic/DyanmicPrefabs/Assets/Scripts/Shared/DynamicPrefabLoadingUtilities.cs

```diff
@@ -0,0 +1,182 @@
+using System.Threading.Tasks;
+using Unity.Netcode;
+using UnityEngine;
+using UnityEngine.AddressableAssets;
+using UnityEngine.ResourceManagement.AsyncOperations;
+
+namespace Game
+{
+    /// <summary>
+    /// A utilities class to handle the loading, tracking, and disposing of loaded network prefabs. Connection and
+    /// disconnection payloads can be easily accessed from this class as well.
+    /// </summary>
+    /// <remarks>
+    /// Artificial delay to the loading of a network prefab is disabled by default. To enable it, make sure to add
+    /// ENABLE_ARTIFICIAL_DELAY as a scripting define symbol to your project's Player settings.
+    /// </remarks>
+    public static class DynamicPrefabLoadingUtilities
+    {
+        const int k_EmptyDynamicPrefabHash = -1;
+
+        static int s_HashOfDynamicPrefabGUIDs = k_EmptyDynamicPrefabHash;
+
+        static Dictionary<AddressableGUID, AsyncOperationHandle<GameObject>> s_LoadedDynamicPrefabResourceHandles = new Dictionary<AddressableGUID, AsyncOperationHandle<GameObject>>();
+        
+        static List<AddressableGUID> s_DynamicPrefabGUIDs = new List<AddressableGUID>(); //cached list to avoid GC
+
+        //A storage where we keep the association between the dynamic prefab (hash of it's GUID) and the clients that have it loaded
+        static Dictionary<int, HashSet<ulong>> s_PrefabHashToClientIds = new Dictionary<int, HashSet<ulong>>();
+
+        public static bool HasClientLoadedPrefab(ulong clientId, int prefabHash) => 
+            s_PrefabHashToClientIds.TryGetValue(prefabHash, out var clientIds) && clientIds.Contains(clientId);
+
+        public static bool IsPrefabLoadedOnAllClients(AddressableGUID assetGuid) => 
+            s_LoadedDynamicPrefabResourceHandles.ContainsKey(assetGuid);
+
+        public static int LoadedPrefabCount => s_LoadedDynamicPrefabResourceHandles.Count;
+
+        public static int ServerPrefabHash => s_HashOfDynamicPrefabGUIDs;
+
+        static NetworkManager s_NetworkManager;
+
+        static DynamicPrefabLoadingUtilities() { }
+
+        public static void Init(NetworkManager networkManager)
+        {
+            s_NetworkManager = networkManager;
+        }
+
+        public static void RecordThatClientHasLoadedAllPrefabs(ulong clientId)
+        {
+            foreach (var dynamicPrefabGUID in s_DynamicPrefabGUIDs)
+            {
+                RecordThatClientHasLoadedAPrefab(dynamicPrefabGUID.GetHashCode(), clientId);
+            }
+        }
+        
+        public static void RecordThatClientHasLoadedAPrefab(int assetGuidHash, ulong clientId)
+        {
+            if (s_PrefabHashToClientIds.TryGetValue(assetGuidHash, out var clientIds))
+            {
+                clientIds.Add(clientId);
+            }
+            else
+            {
+                s_PrefabHashToClientIds.Add(assetGuidHash, new HashSet<ulong>() { clientId });
+            }
+        }
+
+        public static byte[] GenerateRequestPayload()
```

Comment:
not sure if here or somewhere else, but we should comment somewhere about the size limits you identified in your tests.